### PR TITLE
Make time evolution solvers compatible with automatic differentiation

### PR DESCRIPTION
**Pull request by albertomercurio**
## Checklist
Thank you for contributing to `QuantumToolbox.jl`! Please make sure you have finished the following tasks before opening the PR.

- [x] Please read [Contributing to QuantumToolbox.jl](https://qutip.org/QuantumToolbox.jl/stable/resources/contributing).
- [x] Any code changes were done in a way that does not break public API.
- [x] Appropriate tests were added and tested locally by running: `make test`.
- [x] Any code changes should be `julia` formatted by running: `make format`.
- [x] All documents (in `docs/` folder) related to code changes were updated and able to build locally by running: `make docs`.
- [x] (If necessary) the `CHANGELOG.md` should be updated (regarding to the code changes) and built by running: `make changelog`.

Request for a review after you have completed all the tasks. If you have not finished them all, you can also open a [Draft Pull Request](https://github.blog/2019-02-14-introducing-draft-pull-requests/) to let the others know this on-going work.

## Description
With this PR I change the structure of the time evolution solver in order to support automatic differentiation.

Thanks to the [SciMLSensitivity.jl](https://github.com/SciML/SciMLSensitivity.jl) package, it is possible to compute the gradient of a differential equation. It is almost straightforward to do with ODE parameters as a `Vector` type, but it is not easy to implement when we have a complicated structure of the parameters as in the current case of the package, where we have many variables, progress bar, etc inside the params.

The main change here is to introduce a new struct for the parameters, instead of using the current `NamedTuple`. In this way, thanks to [SciMLStructures.jl](https://github.com/SciML/SciMLStructures.jl), we can say which part of the structure is differentiable and which not.

As a first step, I'm trying to simplify the structure of the `params` struct. This involves the creation of a custom struct to handle the `ODEProblem` generated by functions like `sesolveProblem`. In this way, many variables can be removed from `params`.

Currently, there are some limitations on the type of the differentiable part of the `params` struct, and the only supported type is the `Vector` one. For example, the `params` kwarg in the `mesolve` has to be a `Vector` and not a `NamedTuple`. See [this issue](https://github.com/SciML/SciMLSensitivity.jl/issues/1146) for more information. Nonetheless, the `NamedTuple` type is still supported in standard simulations, where the gradient is not needed.

### EDIT:

The custom struct `TimeEvolutionParameters` is no longer needed. We can pass all the cache and temporary variables to their respective callbacks. In this way, the `params` variable is only composed by the **true** parameters (usually a `Vector{Number}`.

### To Do:

- [x] Implement a custom struct for the ODE Parameters
- [x] Implement a custom struct for the `ODEProblem`s generated by the solvers
- [x] Implement the `SciMLStructures.jl` rules for the custom `params` struct
- [x] Make `sesolve` differentiable
- [x] Make `mesolve` differentiable
- [ ] Make `mcsolve` differentiable (maybe in another PR?)

### CODE REVIEW

**Comment by ytdHuang** [comment]
I think for some of the parameters which must be used in all solvers (e.g., `tlist` and `progress_bar`) can still be a independent field in.

**Comment by albertomercurio** [comment]
Actually I'm trying to simplify it as much as possible. The gradient calculation of `sesolve` still complaints about the `ProgressBar`. So I think that I will try to remove it in the next commits to see if this fixes the problem. I'm basically trying to leave inside the `params` struct only `AbstractArray`s.

**Comment by albertomercurio** [comment]
With the last commit, I finally succeeded to support automatic differentiation on `sesolve`. The following code works

```julia
using QuantumToolbox
using OrdinaryDiffEq
using SciMLSensitivity
using Zygote

##

const N = 20
const F = 1
const γ = 1

const a = destroy(N)

const ψ0 = fock(N, 0)

# coef1(p, t) = p.Δ
coef1(p, t) = p[1]

QobjEvo(a' * a, coef1) + F * (a' + a)

function ss_population(Δ)
  # H = Δ * a' * a + F * (a' + a)
  H = QobjEvo(a' * a, coef1) + F * (a' + a)
  c_ops = [sqrt(γ) * a]
  
  tlist = range(0, 1.5, 100)

  ρ_ss = sesolve(H, ψ0, tlist, progress_bar=Val(false), inplace=Val(false), params = [Δ], saveat = [tlist[end]]).states[end].data

  return real(sum(ρ_ss))
end

Δ = 1.0
ss_population(Δ)

##

Zygote.gradient(ss_population, Δ)
```

File: src/qobj/quantum_object_evo.jl [hunk 1]
```diff
@@ -269,7 +271,7 @@ Parse the `op_func_list` and generate the data for the `QuantumObjectEvolution`
     quote
         dims = tuple($(dims_expr...))
 
-        length(unique(dims)) == 1 || throw(ArgumentError("The dimensions of the operators must be the same."))
+        _all_equal(dims) || throw(ArgumentError("The dimensions of the operators must be the same."))
```
[the remark this answers is not in the document]
Comment (albertomercurio) [reply src/qobj/quantum_object_evo.jl]: Oh, good to know.

File: src/time_evolution/callback_helpers/callback_helpers.jl [hunk 2]
```diff
@@ -0,0 +1,93 @@
+#=
+This file contains helper functions for callbacks. The affect! function are defined taking advantage of the Julia struct, which allows to store some cache exclusively for the callback.
+=#
+
+include("sesolve_callback_helpers.jl")
+include("mesolve_callback_helpers.jl")
+include("mcsolve_callback_helpers.jl")
```
Comment (ytdHuang) [inline src/time_evolution/callback_helpers/callback_helpers.jl:7]: I prefer to gather all the `include` calls in `src/QuantumToolbox.jl`

**Review by ytdHuang — APPROVED** [review]

**Pull request merged**